File: android/config-qr-generator/scrambler.js
/*
 * scrambler.js — Algorithm v1.2
 *
 * Port of Scrambler.java / scrambler.cpp.
 * Wire-compatible: output from any implementation decodes correctly in the others.
 *
 * encode(payload: Uint8Array, key: string) → Uint8Array   (length + SALT_LEN)
 * decode(payload: Uint8Array, key: string) → Uint8Array | payload
 *
 * Empty-payload contract (mirrors C++ / Java):
 *   encode(empty) → empty (no salt written)
 *   decode(length ≤ SALT_LEN) → payload unchanged
 */

const SALT_LEN = 4;

// ── private ───────────────────────────────────────────────────────────────────

function _fnv1a(keyBytes) {
    let h = 2166136261; // 0x811C9DC5 — FNV offset basis (unsigned 32-bit)
    for (const b of keyBytes) {
        h = Math.imul(h ^ b, 16777619) >>> 0;
    }
    return h;
}

// xorshift32 — state is Uint32Array[1] (mutated in-place).
// Returns keystreamByte: bits [15:8] of the new state (matches C++ uint8_t cast).
function _xorStep(s) {
    s[0] ^= s[0] << 13;
    s[0] ^= s[0] >>> 17;
    s[0] ^= s[0] << 5;
    return (s[0] >>> 8) & 0xFF;
}

function _initState(key, salt) {
    const keyBytes = new TextEncoder().encode(key || '');
    const seed = (_fnv1a(keyBytes) ^ (Math.imul(salt >>> 0, 0x9E3779B9) >>> 0)) >>> 0;
    const state = new Uint32Array(1);
    state[0] = seed;
    return { state, keyBytes };
}

// ── public API ────────────────────────────────────────────────────────────────

function scramblerEncode(payload, key) {
    if (!payload || payload.length === 0) return payload;

    const saltBytes = new Uint8Array(SALT_LEN);
    crypto.getRandomValues(saltBytes);
    const salt = ((saltBytes[0] << 24) | (saltBytes[1] << 16) |
                  (saltBytes[2] <<  8) |  saltBytes[3]) >>> 0;

    const { state, keyBytes } = _initState(key, salt);
    let prev = salt & 0xFF;

    const out = new Uint8Array(payload.length + SALT_LEN);
    out.set(saltBytes, 0);

    for (let i = 0; i < payload.length; i++) {
        const k   = _xorStep(state);
        let   enc = (payload[i] + prev + k) & 0xFF;
        if (keyBytes.length > 0)
            enc ^= keyBytes[(i + (k & 0x0F)) % keyBytes.length];
        out[i + SALT_LEN] = enc;
        prev = (enc ^ k) & 0xFF;
    }
    return out;
}

function scramblerDecode(payload, key) {
    if (!payload || payload.length <= SALT_LEN) return payload;

    const salt = ((payload[0] << 24) | (payload[1] << 16) |
                  (payload[2] <<  8) |  payload[3]) >>> 0;

    const { state, keyBytes } = _initState(key, salt);
    let prev   = salt & 0xFF;
    const out  = new Uint8Array(payload.length - SALT_LEN);

    for (let i = 0; i < out.length; i++) {
        const enc = payload[i + SALT_LEN];
        const k   = _xorStep(state);
        let   tmp = enc;
        if (keyBytes.length > 0)
            tmp ^= keyBytes[(i + (k & 0x0F)) % keyBytes.length];
        out[i] = (tmp - prev - k) & 0xFF;
        prev   = (enc ^ k) & 0xFF;
    }
    return out;
}

File: esp32/vger32/src/scrambler/scrambler.cpp
/*
 * scrambler.cpp
 *
 * Algorithm v1.2 — per byte:
 *
 *   encode: k    = (uint8_t)(xorshift32(seed) >> 8)
 *           enc  = (in[i] + prev + k) ^ key[(i + (k & 0x0F)) % key_len]
 *           prev = enc ^ k
 *
 *   decode: k    = (uint8_t)(xorshift32(seed) >> 8)   // same PRNG state as encode
 *           temp = enc ^ key[(i + (k & 0x0F)) % key_len]
 *           out  = temp - prev - k
 *           prev = enc ^ k                             // same feedback as encode
 *
 * Salt: 4 random bytes from hardware RNG, prepended to the output.
 * Output length = input length + SCRAMBLER_SALT_LEN, or 0 for empty input.
 * Operates on raw bytes (0x00–0xFF).
 *
 * Changes from v1.1:
 *   - LCG replaced by xorshift32 (passes statistical tests LCG fails)
 *   - Key index shifted by (k & 0x0F) — eliminates key-period attack surface
 *   - prev = enc ^ k — stronger dependency chain between bytes
 */

#include <cstring>
#include <esp_random.h>

#include "scrambler.h"

// ==========================================
// PRIVATE
// ==========================================

static uint32_t hash_key(const char *key) {
    uint32_t h = 2166136261u; // FNV-1a offset basis
    while (*key) {
        h ^= (uint8_t) *key++;
        h *= 16777619u;
    }
    return h;
}

// xorshift32 — better than LCG for all bit positions.
// Returns upper 24 bits (lowest byte is weakest in xorshift32).
static uint32_t prng(uint32_t *s) {
    *s ^= *s << 13;
    *s ^= *s >> 17;
    *s ^= *s << 5;
    return *s >> 8;
}

// ==========================================
// PUBLIC API
// ==========================================

int scrambler_encode(const uint8_t *in, int len, uint8_t *out, const char *key) {
    // Empty payload — no content to scramble, no salt written.
    // Mirrors Scrambler.java behaviour: empty in → empty out.
    if (len == 0) return 0;

    const int key_len = key ? (int) strlen(key) : 0;

    uint32_t salt = esp_random();
    out[0] = (uint8_t) ((salt >> 24) & 0xFF);
    out[1] = (uint8_t) ((salt >> 16) & 0xFF);
    out[2] = (uint8_t) ((salt >> 8) & 0xFF);
    out[3] = (uint8_t) (salt & 0xFF);

    uint32_t seed = hash_key(key ? key : "") ^ (salt * 0x9E3779B9u);
    uint8_t prev = (uint8_t) (salt & 0xFF);

    for (int i = 0; i < len; i++) {
        uint8_t k = (uint8_t) prng(&seed);
        uint8_t enc = (uint8_t) (in[i] + prev + k);
        if (key_len > 0)
            enc ^= (uint8_t) key[(i + (k & 0x0F)) % key_len];
        out[i + SCRAMBLER_SALT_LEN] = enc;
        prev = enc ^ k;
    }
    return len + SCRAMBLER_SALT_LEN;
}

int scrambler_decode(const uint8_t *in, int len, uint8_t *out, const char *key) {
    if (len <= SCRAMBLER_SALT_LEN) return 0;

    const int key_len = key ? (int) strlen(key) : 0;

    uint32_t salt = ((uint32_t) in[0] << 24)
                    | ((uint32_t) in[1] << 16)
                    | ((uint32_t) in[2] << 8)
                    | (uint32_t) in[3];

    uint32_t seed = hash_key(key ? key : "") ^ (salt * 0x9E3779B9u);
    uint8_t prev = (uint8_t) (salt & 0xFF);
    int out_len = len - SCRAMBLER_SALT_LEN;

    for (int i = 0; i < out_len; i++) {
        uint8_t enc = in[i + SCRAMBLER_SALT_LEN];
        uint8_t k = (uint8_t) prng(&seed);
        uint8_t temp = enc;
        if (key_len > 0)
            temp ^= (uint8_t) key[(i + (k & 0x0F)) % key_len];
        out[i] = (uint8_t) (temp - prev - k);
        prev = enc ^ k;
    }
    return out_len;
}
